enhancements/is-bundle-proxying.md

```diff
@@ -0,0 +1,81 @@
+- [ ] Graduation criteria for dev preview, tech preview, GA
+
+## Open Questions
+
+- External images can be pulled into OpenShift’s integrated image registry
+- Pulling an external image to the integrated registry can be triggered by creating/updating an ImageStream resource
+- No non-standard cluster configuration is needed to allow external images to be pulled into the integrated registry
+- Pulls of external images into the registry abide by ImageContentSourcePolicy and other applicable cluster configurations
+- Direct -- non-pod spec -- pulling of images from the integrated registry transitively satisfies ImageContentSourcePolicy constraints
+
+## Summary
+
+When deployed to OpenShift, OLM can use an ImageStream to stage a bundle image in the internal image registry and access its image metadata while still meeting the OpenShift-specific constraint to pull images via CRI-O.
+
+## Motivation
+
+Today, when a bundle image is pulled and unpacked by OLM, a Job is created which specifies three containers:
+
+- a bundle util init container
+  - contains a statically linked bundle tool that can search the container’s filesystem for bundle metadata -- annotations.yaml -- and copy the referenced manifest files to a bundle volume
+  - mounts a single bundle util volume
+  - at runtime, uses cp to copy this bundle tool to the bundle util volume
+- a bundle data init container
+  - pull spec is used to delegate pulling of the bundle image via CRI-O
+  - mounts two volumes: bundle util and bundle data
+  - at runtime, executes the bundle tool from the bundle util mount to copy the bundle to the bundle data mount
+- a bundle extract container
+  - contains opm
+  - mounts the bundle data volume
+  - runs `opm extract` against the bundle data mount which results in a ConfigMap being created that contains the bundle’s kubectl-able content
+
+While this works for both OpenShift and upstream Kubernetes, the inability to inspect the image annotations -- where the path of the metadata directory is specified -- means that we cannot verify that the bundle metadata directory found in the bundle’s filesystem contains the data intended by the author.
+
+Using opm to directly interact with the image, instead of relying on a pod pull spec, allows for unambiguous resolution of the bundle metadata directory. In the future, this will also enable bundle data to be stored entirely as image annotations and/or manifest lists (w/o a run configuration or union filesystem).
+
+### Goals
+
+- Disambiguate the result of unpacking a bundle on-cluster
+- When deployed to OpenShift, adhere to constraints imposed by ImageContentSourcePolicy when pulling bundle images
+
+### Non-Goals
+
+- Make ImageStreams a hard requirement for bundle unpacking on platforms other than OpenShift
+
+## Proposal
+
+### Implementation Details/Notes/Constraints
+
+To unpack a bundle image, OLM will generate a job:
+
+- with a single container spec
+- that uses opm to directly pull the image (daemonless)
```

Comment:
so opm, running inside this container, is going to respect:

cluster proxy settings+CAs
cluster imagecontentsourcepolicy
namespace/serviceaccount/node pull secrets
cluster image policy (allowlist/blocklist, trusted/insecure registry definitions)
cluster defined external registry CAs

and all the other behaviors such that opm pulling an image behaves identically to the experience of a node pulling an image on behalf of a pod?

This is a lot like the work that builds took on, and it's been a hard struggle to get there/stay there.  I believe even today there are gaps where a build pulling an image does not behave identically to a pod pulling an image, and it's always a source of confusion/frustration.

Reply:
This proposal is trying to _make use_ of the work for builds instead of reimplementing it. The assumption is that we get these behaviors for free by creating an ImageStream and referencing the integrated registry (still an open question, which is why this is provisional).

Reply:
yes you can avoid most of these problems by using pullthrough against the internal registry.  but now you're dependent on the internal registry being enabled/running on all clusters.

Reply:
> but now you're dependent on the internal registry being enabled/running on all clusters.

So it sounds like a registry-less configuration is common enough to be concerned about it. Is that correct?

Reply:
> So it sounds like a registry-less configuration is common enough to be concerned about it. Is that correct?

yeah.  
cc @jwforres

Reply:
We should avoid technical decisions that would tie us to always having the registry running on a cluster. I think it is safe to assume the existence of ImageStreams, but not necessarily the integrated registry.